bluemira/magnetostatics/circuits.py

```diff
@@ -37,80 +37,21 @@
 )
 from bluemira.magnetostatics.baseclass import CurrentSource, SourceGroup
 from bluemira.magnetostatics.error import MagnetostaticsError
+from bluemira.magnetostatics.polyhedral_prism import PolyhedralPrismCurrentSource
 from bluemira.magnetostatics.tools import process_to_coordinates, process_xyz_array
 from bluemira.magnetostatics.trapezoidal_prism import TrapezoidalPrismCurrentSource
 
 __all__ = ["ArbitraryPlanarRectangularXSCircuit", "HelmholtzCage"]
 
 
-class ArbitraryPlanarRectangularXSCircuit(SourceGroup):
+class PlanarCircuit(SourceGroup):
     """
-    An arbitrary, planar current loop of constant rectangular cross-section
-    and uniform current density.
-
-    Parameters
-    ----------
-    shape:
-        The geometry from which to form an ArbitraryPlanarRectangularXSCircuit
-    breadth:
-        The breadth of the current source (half-width) [m]
-    depth:
-        The depth of the current source (half-height) [m]
-    current:
-        The current flowing through the source [A]
-
-    Notes
-    -----
-    Works best with planar x-z geometries.
+    Base class for a planar current loop
     """
 
     shape: np.array
-    breadth: float
-    depth: float
     current: float
 
-    def __init__(
-        self,
-        shape: Union[np.ndarray, Coordinates],
-        breadth: float,
-        depth: float,
-        current: float,
-    ):
-        shape = process_to_coordinates(shape)
-        if not shape.is_planar:
-            raise MagnetostaticsError(
-                f"The input shape for {self.__class__.__name__} must be planar."
-            )
-
-        betas, alphas = self._get_betas_alphas(shape)
-
-        normal = shape.normal_vector
-
-        # Set up geometry, calculating all trapezoidal prism sources
-        self.shape = shape.T
-        self.d_l = np.diff(self.shape, axis=0)
-        self.midpoints = self.shape[:-1, :] + 0.5 * self.d_l
-        sources = []
-
-        for midpoint, d_l, beta, alpha in zip(self.midpoints, self.d_l, betas, alphas):
-            d_l_norm = d_l / np.linalg.norm(d_l)
-            t_vec = np.cross(d_l_norm, normal)
-
-            source = TrapezoidalPrismCurrentSource(
-                midpoint,
-                d_l,
-                normal,
-                t_vec,
-                breadth,
-                depth,
-                alpha,
-                beta,
-                current,
-            )
-            sources.append(source)
-
-        super().__init__(sources)
-
     def _get_betas_alphas(self, shape):
```

Comment:
sonarcloud is moaning about duplication so:
```suggestion
    def _initialise_source_mechanics(self, shape, current):
        shape = process_to_coordinates(shape)
        if not shape.is_planar:
            raise MagnetostaticsError(
                f"The input shape for {type(self).__name__} must be planar."
            )

        betas, alphas = self._get_betas_alphas(shape)

        normal = shape.normal_vector

        # Set up geometry, calculating all sources
        self.shape = shape.T
        self.d_l = np.diff(self.shape, axis=0)
        self.midpoints = self.shape[:-1, :] + 0.5 * self.d_l

        for midpoint, d_l, beta, alpha in zip(self.midpoints, self.d_l, betas, alphas):
            d_l_norm = d_l / np.linalg.norm(d_l)
            t_vec = np.cross(d_l_norm, normal)
            yield midpoint, d_l, normal, t_vec, alpha, beta

    def _get_betas_alphas(self, shape):
```

then in the child functions eg Trapezoidal:
```python
sources = [TrapezoidalPrismCurrentSource(
                midpoint,
                d_l,
                normal,
                t_vec,
                breadth,
                depth,
                alpha,
                beta,
                current,
            ) for midpoint, d_l, normal, t_vec, alpha, beta in self._initialise_source_mechanics(shape, current)]

super().__init__(sources)
````